File: src/my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool

class NumberCounter(Node):
  def __init__(self):
    super().__init__("number_counter")
    self.count_ = 0
    self.number_subscriber_ = self.create_subscription(Int64, "number", self.count_number_callback, 10)
    self.number_count_publisher_ = self.create_publisher(Int64, "number_count", 10)
    self.server_ = self.create_service(SetBool, "reset_counter", self.reset_counter_callback)
    self.get_logger().info("NumberCounter node has been started.")

  def count_number_callback(self, msg: Int64):
    self.count_ += msg.data
    count_msg = Int64()
    count_msg.data = self.count_
    self.number_count_publisher_.publish(count_msg)
  
  def reset_counter_callback(self, request: SetBool.Request, response: SetBool.Response):
    if (request.data):
      self.count_ = 0
      response.success = True
      response.message = "Counter reset"
    else:
      response.message = False
      response.message = "Counter not reset"
    return response

def main(args=None):
  rclpy.init(args=args)
  node = NumberCounter()
  rclpy.spin(node)
  rclpy.shutdown()


if __name__ == "__main__":
  main()

File: src/my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

using namespace std::placeholders;

class NumberCounter : public rclcpp::Node
{
public:
  NumberCounter() : Node("number_counter")
  {
    number_subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
        "number", 10,
        std::bind(&NumberCounter::timerCallbackCountNumber, this, _1)
    );
    number_count_publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);
    reset_counter_server_ = this->create_service<example_interfaces::srv::SetBool>(
        "reset_counter",
        std::bind(&NumberCounter::resetCounterCallback, this, _1, _2)
    );
    RCLCPP_INFO(this->get_logger(), "NumberCounter node has been started.");
  }

private:
  void timerCallbackCountNumber(const example_interfaces::msg::Int64::SharedPtr msg)
  {
    count_ += msg->data;
    auto count_msg = example_interfaces::msg::Int64();
    count_msg.data = count_;
    number_count_publisher_->publish(count_msg);
  }

  void resetCounterCallback(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                            const example_interfaces::srv::SetBool::Response::SharedPtr response)
  {
    if (request->data) {
      count_ = 0;
      response->success = true;
      response->message = "Counter reset";
    } else {
      response->success = false;
      response->message =  "Counter not reset";
    }
  }

  int64_t count_;
  rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr number_count_publisher_;
  rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr number_subscriber_;
  rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr reset_counter_server_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<NumberCounter>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}